lib/uplinkc/access_scope.go

```diff
@@ -101,6 +103,55 @@ func serialize_scope(scopeRef C.ScopeRef, cerr **C.char) *C.char {
 	return C.CString(serializedScope)
 }
 
+//export restrict_scope
+// restrict_scope restricts a given scope with the provided caveat and encryption restrictions
+func restrict_scope(scopeRef C.ScopeRef, caveat C.Caveat, restrictions **C.EncryptionRestriction, numberOfRestrictions int, cerr **C.char) C.ScopeRef {
+	scope, ok := universe.Get(scopeRef._handle).(*libuplink.Scope)
+	if !ok {
+		*cerr = C.CString("invalid scope")
+		return C.ScopeRef{}
+	}
+
+	caveatGo := macaroon.Caveat{
+		DisallowReads:   caveat.disallow_reads == C.bool(true),
+		DisallowWrites:  caveat.disallow_writes == C.bool(true),
+		DisallowLists:   caveat.disallow_lists == C.bool(true),
+		DisallowDeletes: caveat.disallow_deletes == C.bool(true),
+	}
+
+	apiKeyRestricted, err := scope.APIKey.Restrict(caveatGo)
+	if err != nil {
+		*cerr = C.CString(fmt.Sprintf("%+v", err))
+		return C.ScopeRef{}
+	}
+
+	restrictionsGo := make([]libuplink.EncryptionRestriction, 0, numberOfRestrictions)
+	if restrictions != nil {
+		restrictionsArray := (*[1 << 30 / unsafe.Sizeof(C.EncryptionRestriction{})]C.EncryptionRestriction)(unsafe.Pointer(restrictions))
```

Comment:
I don't get what `1 << 30 / unsafe.Sizeof(C.EncryptionRestriction{})` value stands for.

Reply:
I'm no expert here and I just adjusted code found in the example. My understanding is that is line is mapping pure pointer to an array with a constant length and `1 << 30 / unsafe.Sizeof(C.EncryptionRestriction{})` is length of array large enough to keep all potential elements.

Reply:
I'm not an expert either but it's confusing, so we should drop a comment about it.

On the other hand, my rationale is that this array should be of the `numberOfRestrictions * unsafe.Sizeof(C.EncryptionRestriction{})` however I cannot make that change and test it because I've never been able to run the `libuplinkc` test in my local (I always get compilation errors https://storjlabs.atlassian.net/browse/V3-2099).

Could you give it a try and if breaks, then drop a comment?
We could also mention some specific people that they have more expertise in the field to check this line.

Reply:
Expression `numberOfRestrictions * unsafe.Sizeof(C.EncryptionRestriction{})` wont work because its not constant value. What I can do it to add a comment.

Regarding double-check, mainly @bryanchriswhite and @egonelbre were involved in `uplinkc`.

Reply:
I see.
I'd prefer to use a slice if it's possible; I've seen that there is some way to initialize a slice and access to the underlying array, but maybe it's too hacky.

Reply:
`1 << 32 / unsafe.Sizeof(C.EncryptionRestriction{})` is the largest slice you can have of EncryptionRestriction on 32 bit platform. `1 << 30 / unsafe.Sizeof(C.EncryptionRestriction{})` is smaller so the end pointer wouldn't overflow address space.

Reply:
Let's add a comment, it isn't obvious

Reply:
AFAIR there's an issue with that expression that needs to be fixed anyways, but I cannot recall the exact solution at the moment.

Reply:
Is there a ticket? otherwise, we could open one.
I would to the comment a `TODO: v3-XXXX`

Reply:
Ah, found it:

```
*(*[]byte)(unsafe.Pointer(
        &reflect.SliceHeader{
            Data: uintptr(p),
            Len:  int(size),
            Cap:  int(size),
        },
    ),
)
```

The main reason people ended-up using `1<<30` thing is because it was shorter to write, but with upcoming `checkptr` it won't work.

Reply:
See more in https://blog.gopheracademy.com/advent-2019/safe-use-of-unsafe-pointer/

Reply:
https://storjlabs.atlassian.net/browse/V3-3351

Reply:
`reflect.SliceHeader` looks to be working, thanks @egonelbre 

@ifraixedes let me know if you think we still need commenting this part.

Reply:
Also created https://review.dev.storj.io/c/storj/storj/+/246

Reply:
no, there is no need with the change.